File: packages/essential_sdk_dart/lib/src/essential_types.dart
import 'dart:io';

import 'essential_llama_runtime.dart';

enum EssentialErrorCode {
  modelNotInstalled('MODEL_NOT_INSTALLED'),
  modelIncompatible('MODEL_INCOMPATIBLE'),
  adapterIncompatible('ADAPTER_INCOMPATIBLE'),
  deviceCapacityInsufficient('DEVICE_CAPACITY_INSUFFICIENT'),
  permissionDenied('PERMISSION_DENIED'),
  sessionCancelled('SESSION_CANCELLED'),
  runtimeUnavailable('RUNTIME_UNAVAILABLE'),
  invalidConfiguration('INVALID_CONFIGURATION'),
  requestTimedOut('REQUEST_TIMED_OUT'),
  unsupportedTaskType('UNSUPPORTED_TASK_TYPE'),
  payloadSchemaInvalid('PAYLOAD_SCHEMA_INVALID'),
  modalityPermissionDenied('MODALITY_PERMISSION_DENIED'),
  capabilityNotAvailable('CAPABILITY_NOT_AVAILABLE'),
  bundleDependencyMissing('BUNDLE_DEPENDENCY_MISSING'),
  realtimeNotSupported('REALTIME_NOT_SUPPORTED');

  const EssentialErrorCode(this.wireName);

  final String wireName;
}

class EssentialException implements Exception {
  const EssentialException(this.code, this.message, {this.cause});

  final EssentialErrorCode code;
  final String message;
  final Object? cause;

  @override
  String toString() => 'EssentialException(${code.wireName}): $message';
}

class EssentialInstalledModel {
  const EssentialInstalledModel({
    required this.modelId,
    required this.modelPath,
    this.family,
    this.capabilities = const <String>[],
    this.contextWindow,
    this.supportsAdapters = false,
  });

  final String modelId;
  final String modelPath;
  final String? family;
  final List<String> capabilities;
  final int? contextWindow;
  final bool supportsAdapters;

  bool get isInstalled => File(modelPath).existsSync();
}

class EssentialModelRequirement {
  const EssentialModelRequirement({
    this.modelId,
    this.family,
    this.capability,
    this.minContextWindow,
    this.maxLatencyMs,
    this.allowFallback = true,
    this.explicitModelPath,
  });

  const EssentialModelRequirement.anyCompatible({
    this.family,
    this.capability,
    this.minContextWindow,
    this.maxLatencyMs,
  }) : modelId = null,
       allowFallback = true,
       explicitModelPath = null;

  const EssentialModelRequirement.fixed(
    String this.modelId, {
    this.family,
    this.capability,
    this.minContextWindow,
    this.maxLatencyMs,
  }) : allowFallback = false,
       explicitModelPath = null;

  const EssentialModelRequirement.fallback(
    String this.modelId, {
    this.family,
    this.capability,
    this.minContextWindow,
    this.maxLatencyMs,
  }) : allowFallback = true,
       explicitModelPath = null;

  final String? modelId;
  final String? family;
  final String? capability;
  final int? minContextWindow;
  final int? maxLatencyMs;
  final bool allowFallback;
  final String? explicitModelPath;

  EssentialModelRequirement copyWith({
    String? modelId,
    String? family,
    String? capability,
    int? minContextWindow,
    int? maxLatencyMs,
    bool? allowFallback,
    String? explicitModelPath,
  }) {
    return EssentialModelRequirement(
      modelId: modelId ?? this.modelId,
      family: family ?? this.family,
      capability: capability ?? this.capability,
      minContextWindow: minContextWindow ?? this.minContextWindow,
      maxLatencyMs: maxLatencyMs ?? this.maxLatencyMs,
      allowFallback: allowFallback ?? this.allowFallback,
      explicitModelPath: explicitModelPath ?? this.explicitModelPath,
    );
  }
}

enum EssentialInputKind { text, image, audio, url, document, reference }

class EssentialInputAttachment {
  const EssentialInputAttachment({
    required this.kind,
    this.text,
    this.filePath,
    this.uri,
    this.mimeType,
    this.name,
    this.metadata = const <String, Object?>{},
  });

  const EssentialInputAttachment.text(
    String this.text, {
    this.name,
    this.metadata = const <String, Object?>{},
  }) : kind = EssentialInputKind.text,
       filePath = null,
       uri = null,
       mimeType = 'text/plain';

  const EssentialInputAttachment.imageFile(
    String this.filePath, {
    this.name,
    this.mimeType = 'image/*',
    this.metadata = const <String, Object?>{},
  }) : kind = EssentialInputKind.image,
       text = null,
       uri = null;

  const EssentialInputAttachment.audioFile(
    String this.filePath, {
    this.name,
    this.mimeType = 'audio/*',
    this.metadata = const <String, Object?>{},
  }) : kind = EssentialInputKind.audio,
       text = null,
       uri = null;

  const EssentialInputAttachment.url(
    String this.uri, {
    this.name,
    this.metadata = const <String, Object?>{},
  }) : kind = EssentialInputKind.url,
       text = null,
       filePath = null,
       mimeType = 'text/uri-list';

  final EssentialInputKind kind;
  final String? text;
  final String? filePath;
  final String? uri;
  final String? mimeType;
  final String? name;
  final Map<String, Object?> metadata;
}

class EssentialReferenceDocument {
  const EssentialReferenceDocument({
    required this.title,
    this.text = '',
    this.uri,
    this.filePath,
    this.mimeType = 'text/plain',
    this.metadata = const <String, Object?>{},
  });

  final String title;
  final String text;
  final String? uri;
  final String? filePath;
  final String mimeType;
  final Map<String, Object?> metadata;
}

class EssentialRuntimeOptions {
  const EssentialRuntimeOptions({
    this.preferredModelId,
    this.webSearchEnabled = false,
    this.locationEnabled = false,
    this.sharedMemoryEnabled = false,
    bool? sharedMemoryReadEnabled,
    bool? sharedMemoryWriteEnabled,
    this.spokenOutputEnabled = false,
  }) : sharedMemoryReadEnabled = sharedMemoryReadEnabled ?? sharedMemoryEnabled,
       sharedMemoryWriteEnabled =
           sharedMemoryWriteEnabled ?? sharedMemoryEnabled;

  final String? preferredModelId;
  final bool webSearchEnabled;
  final bool locationEnabled;
  @Deprecated(
    'Use sharedMemoryReadEnabled and sharedMemoryWriteEnabled for per-request control.',
  )
  final bool sharedMemoryEnabled;
  final bool sharedMemoryReadEnabled;
  final bool sharedMemoryWriteEnabled;
  final bool spokenOutputEnabled;
}

class EssentialGenerateRequest {
  const EssentialGenerateRequest({
    this.id,
    this.sessionId,
    required this.prompt,
    this.systemInstruction = '',
    this.attachments = const <EssentialInputAttachment>[],
    this.referenceDocuments = const <EssentialReferenceDocument>[],
    this.modelRequirement = const EssentialModelRequirement.anyCompatible(),
    this.runtimeOptions = const EssentialRuntimeOptions(),
    this.maxTokens = 64,
    this.topK = 40,
    this.topP = 0.95,
    this.temperature = 0.8,
    this.seed = 42,
    this.timeoutMs,
  });

  final String? id;
  final String? sessionId;
  final String prompt;
  final String systemInstruction;
  final List<EssentialInputAttachment> attachments;
  final List<EssentialReferenceDocument> referenceDocuments;
  final EssentialModelRequirement modelRequirement;
  final EssentialRuntimeOptions runtimeOptions;
  final int maxTokens;
  final int topK;
  final double topP;
  final double temperature;
  final int seed;
  final int? timeoutMs;

  EssentialGenerateRequest copyWith({
    String? id,
    String? sessionId,
    String? prompt,
    String? systemInstruction,
    List<EssentialInputAttachment>? attachments,
    List<EssentialReferenceDocument>? referenceDocuments,
    EssentialModelRequirement? modelRequirement,
    EssentialRuntimeOptions? runtimeOptions,
    int? maxTokens,
    int? topK,
    double? topP,
    double? temperature,
    int? seed,
    int? timeoutMs,
  }) {
    return EssentialGenerateRequest(
      id: id ?? this.id,
      sessionId: sessionId ?? this.sessionId,
      prompt: prompt ?? this.prompt,
      systemInstruction: systemInstruction ?? this.systemInstruction,
      attachments: attachments ?? this.attachments,
      referenceDocuments: referenceDocuments ?? this.referenceDocuments,
      modelRequirement: modelRequirement ?? this.modelRequirement,
      runtimeOptions: runtimeOptions ?? this.runtimeOptions,
      maxTokens: maxTokens ?? this.maxTokens,
      topK: topK ?? this.topK,
      topP: topP ?? this.topP,
      temperature: temperature ?? this.temperature,
      seed: seed ?? this.seed,
      timeoutMs: timeoutMs ?? this.timeoutMs,
    );
  }
}

class EssentialGenerateChunk {
  const EssentialGenerateChunk({
    required this.requestId,
    required this.delta,
    required this.accumulatedText,
    required this.modelUsed,
  });

  final String requestId;
  final String delta;
  final String accumulatedText;
  final String modelUsed;
}

class EssentialGenerateResult {
  const EssentialGenerateResult({
    required this.requestId,
    required this.text,
    required this.modelUsed,
    required this.finishReason,
  });

  final String requestId;
  final String text;
  final String modelUsed;
  final String finishReason;
}

class EssentialConfiguration {
  const EssentialConfiguration({
    this.installedModels = const <EssentialInstalledModel>[],
    this.defaultModelId,
    this.modelLoadOptions = const EssentialLlamaModelOptions(),
  });

  final List<EssentialInstalledModel> installedModels;
  final String? defaultModelId;
  final EssentialLlamaModelOptions modelLoadOptions;
}

File: packages/essential_android_sdk/src/main/java/io/essential/sdk/android/EssentialTypes.kt
package io.essential.sdk.android

import java.io.File
import java.util.UUID

data class EssentialServiceConfiguration(
    val servicePackage: String,
    val serviceClassName: String,
    val callerPackage: String? = null,
    val defaultModelId: String? = null,
)

data class EssentialModelRequirement(
    val modelId: String? = null,
    val family: String? = null,
    val capability: String? = null,
    val minContextWindow: Int? = null,
    val maxLatencyMs: Int? = null,
    val allowFallback: Boolean = true,
    val explicitModelPath: String? = null,
) {
    companion object {
        fun anyCompatible(
            family: String? = null,
            capability: String? = null,
            minContextWindow: Int? = null,
            maxLatencyMs: Int? = null,
        ) = EssentialModelRequirement(
            family = family,
            capability = capability,
            minContextWindow = minContextWindow,
            maxLatencyMs = maxLatencyMs,
            allowFallback = true,
        )

        fun fixed(
            modelId: String,
            family: String? = null,
            capability: String? = null,
            minContextWindow: Int? = null,
            maxLatencyMs: Int? = null,
        ) = EssentialModelRequirement(
            modelId = modelId,
            family = family,
            capability = capability,
            minContextWindow = minContextWindow,
            maxLatencyMs = maxLatencyMs,
            allowFallback = false,
        )

        fun fallback(
            preferredModelId: String,
            family: String? = null,
            capability: String? = null,
            minContextWindow: Int? = null,
            maxLatencyMs: Int? = null,
        ) = EssentialModelRequirement(
            modelId = preferredModelId,
            family = family,
            capability = capability,
            minContextWindow = minContextWindow,
            maxLatencyMs = maxLatencyMs,
            allowFallback = true,
        )

        fun explicit(
            modelPath: String,
            modelId: String? = null,
            family: String? = null,
            capability: String? = null,
            minContextWindow: Int? = null,
            maxLatencyMs: Int? = null,
            allowFallback: Boolean = true,
        ) = EssentialModelRequirement(
            modelId = modelId,
            family = family,
            capability = capability,
            minContextWindow = minContextWindow,
            maxLatencyMs = maxLatencyMs,
            allowFallback = allowFallback,
            explicitModelPath = modelPath,
        )
    }
}

data class EssentialGenerateRequest(
    val id: String? = null,
    val sessionId: String? = null,
    val prompt: String,
    val systemInstruction: String = "",
    val attachments: List<EssentialMediaAttachment> = emptyList(),
    val referenceDocuments: List<EssentialReferenceDocument> = emptyList(),
    val modelRequirement: EssentialModelRequirement = EssentialModelRequirement.anyCompatible(),
    val runtimeOptions: EssentialRuntimeOptions = EssentialRuntimeOptions(),
    val adapterId: String? = null,
    val maxTokens: Int = 128,
    val topK: Int = 40,
    val topP: Double = 0.95,
    val temperature: Double = 0.8,
    val seed: Int = 42,
    val timeoutMs: Long? = null,
) {
    fun withRequestId(requestId: String): EssentialGenerateRequest {
        return copy(id = requestId)
    }
}

data class EssentialGenerateChunk(
    val requestId: String,
    val delta: String,
    val accumulatedText: String,
    val modelUsed: String,
)

data class EssentialRuntimeOptions(
    val preferredModelId: String? = null,
    val webSearchEnabled: Boolean = false,
    val locationEnabled: Boolean = false,
    @Deprecated(
        message = "Use sharedMemoryReadEnabled and sharedMemoryWriteEnabled for per-request control.",
        replaceWith = ReplaceWith("sharedMemoryReadEnabled && sharedMemoryWriteEnabled"),
    )
    val sharedMemoryEnabled: Boolean = false,
    val sharedMemoryReadEnabled: Boolean = sharedMemoryEnabled,
    val sharedMemoryWriteEnabled: Boolean = sharedMemoryEnabled,
    val spokenOutputEnabled: Boolean = false,
)

data class EssentialGenerateResult(
    val requestId: String,
    val text: String,
    val modelUsed: String,
    val finishReason: String,
)

enum class EssentialTaskType(val wireName: String) {
    TEXT_GENERATION("text_generation"),
    MULTIMODAL_CHAT("multimodal_chat"),
    IMAGE_CAPTION("image_caption"),
    PLANT_IDENTIFICATION("plant_identification"),
    STT("stt"),
    TTS("tts"),
    VOICE_CONVERSATION("voice_conversation"),
}

enum class EssentialMediaKind(val wireName: String) {
    IMAGE("image"),
    AUDIO("audio"),
    TEXT("text"),
    URL("url"),
    DOCUMENT("document"),
    REFERENCE("reference"),
}

data class EssentialMediaAttachment(
    val kind: EssentialMediaKind,
    val uri: String? = null,
    val filePath: String? = null,
    val mimeType: String? = null,
    val width: Int? = null,
    val height: Int? = null,
    val durationMs: Long? = null,
    val metadata: Map<String, String> = emptyMap(),
)

data class EssentialReferenceDocument(
    val title: String,
    val text: String = "",
    val uri: String? = null,
    val filePath: String? = null,
    val mimeType: String = "text/plain",
    val metadata: Map<String, String> = emptyMap(),
)

data class EssentialTaskRequest(
    val id: String? = null,
    val sessionId: String? = null,
    val taskType: EssentialTaskType,
    val prompt: String? = null,
    val modelRequirement: EssentialModelRequirement = EssentialModelRequirement.anyCompatible(),
    val attachments: List<EssentialMediaAttachment> = emptyList(),
    val referenceDocuments: List<EssentialReferenceDocument> = emptyList(),
    val systemInstruction: String = "",
    val runtimeOptions: EssentialRuntimeOptions = EssentialRuntimeOptions(),
    val maxTokens: Int = 256,
    val topK: Int = 40,
    val topP: Double = 0.95,
    val temperature: Double = 0.7,
    val seed: Int = 42,
    val timeoutMs: Long? = null,
    val metadata: Map<String, String> = emptyMap(),
) {
    fun withRequestId(requestId: String): EssentialTaskRequest = copy(id = requestId)

    companion object {
        fun pixelFeatureChat(
            prompt: String,
            sessionId: String? = null,
            image: EssentialMediaAttachment? = null,
            audio: EssentialMediaAttachment? = null,
            references: List<EssentialReferenceDocument> = emptyList(),
            runtimeOptions: EssentialRuntimeOptions = EssentialRuntimeOptions(
                webSearchEnabled = true,
                locationEnabled = true,
                sharedMemoryReadEnabled = false,
                sharedMemoryWriteEnabled = false,
            ),
            modelRequirement: EssentialModelRequirement = EssentialModelRequirement.anyCompatible(
                capability = "multimodal_chat",
            ),
        ): EssentialTaskRequest = EssentialTaskRequest(
            sessionId = sessionId,
            taskType = EssentialTaskType.MULTIMODAL_CHAT,
            prompt = prompt,
            modelRequirement = modelRequirement,
            attachments = listOfNotNull(image, audio),
            referenceDocuments = references,
            runtimeOptions = runtimeOptions,
            systemInstruction = """
                You are a Pixel feature support assistant running through the Essential AI API.
                Web search and location context may be supplied by the host app; when present, use them directly and do not claim that web search or location access is unavailable.
                Shared memory is optional past context only. If memory is disabled or not relevant, ignore it completely.
                For product URLs, product names, or product photos, identify the item, use web/review context when supplied, and compare options by use case.
                Answer in the user's language with practical steps and concise cautions.
            """.trimIndent(),
            metadata = mapOf(
                "demo" to "pixel_feature_chat",
                "domain" to "pixel_usage_support",
                "web_first" to runtimeOptions.webSearchEnabled.toString(),
                "location_context" to runtimeOptions.locationEnabled.toString(),
                "shared_memory_read" to runtimeOptions.sharedMemoryReadEnabled.toString(),
                "shared_memory_write" to runtimeOptions.sharedMemoryWriteEnabled.toString(),
            ),
        )

        @Deprecated(
            message = "Use pixelFeatureChat for the external AI API demo app.",
            replaceWith = ReplaceWith("pixelFeatureChat(prompt, sessionId, image, audio, references, modelRequirement = modelRequirement)"),
        )
        fun macHelpChat(
            prompt: String,
            sessionId: String? = null,
            image: EssentialMediaAttachment? = null,
            audio: EssentialMediaAttachment? = null,
            references: List<EssentialReferenceDocument> = emptyList(),
            modelRequirement: EssentialModelRequirement = EssentialModelRequirement.anyCompatible(
                capability = "multimodal_chat",
            ),
        ): EssentialTaskRequest = EssentialTaskRequest(
            sessionId = sessionId,
            taskType = EssentialTaskType.MULTIMODAL_CHAT,
            prompt = prompt,
            modelRequirement = modelRequirement,
            attachments = listOfNotNull(image, audio),
            referenceDocuments = references,
            metadata = mapOf(
                "demo" to "mac_help_chat",
                "domain" to "mac_usage_support",
            ),
        )

        fun productQaChat(
            productName: String,
            question: String,
            references: List<EssentialReferenceDocument>,
            images: List<EssentialMediaAttachment> = emptyList(),
            sessionId: String? = null,
            modelRequirement: EssentialModelRequirement = EssentialModelRequirement.anyCompatible(
                capability = "multimodal_chat",
            ),
        ): EssentialTaskRequest = EssentialTaskRequest(
            sessionId = sessionId,
            taskType = EssentialTaskType.MULTIMODAL_CHAT,
            prompt = "商品「$productName」について質問に答えてください。\n質問: $question",
            modelRequirement = modelRequirement,
            attachments = images,
            referenceDocuments = references,
            systemInstruction = "参照資料を最優先し、資料にない内容は推測と明記してください。",
            metadata = mapOf(
                "domain" to "product_qa",
                "productName" to productName,
            ),
        )

        fun plantIdentification(
            image: EssentialMediaAttachment,
            prompt: String = "この植物の名前、特徴、育て方の注意点を日本語で説明してください。",
            modelRequirement: EssentialModelRequirement = EssentialModelRequirement.anyCompatible(
                capability = "plant_identification",
            ),
        ): EssentialTaskRequest = EssentialTaskRequest(
            taskType = EssentialTaskType.PLANT_IDENTIFICATION,
            prompt = prompt,
            modelRequirement = modelRequirement,
            attachments = listOf(image),
            metadata = mapOf("demo" to "plant_identification"),
        )
    }
}

data class EssentialTaskResult(
    val requestId: String,
    val taskType: EssentialTaskType,
    val status: String,
    val text: String,
    val modelUsed: String,
    val finishReason: String,
    val metadata: Map<String, String> = emptyMap(),
)

data class EssentialTaskChunk(
    val requestId: String,
    val taskType: EssentialTaskType,
    val delta: String,
    val accumulatedText: String,
    val modelUsed: String,
)

data class EssentialModelDescriptor(
    val modelId: String,
    val family: String?,
    val path: String?,
    val sizeBytes: Long,
    val route: String,
    val runtimeFamily: String = route,
    val supportsImage: Boolean = false,
    val supportsAudio: Boolean = false,
) {
    val isInstalled: Boolean
        get() = !path.isNullOrBlank() || route == "mock"
}

data class EssentialAdapterDescriptor(
    val adapterId: String,
    val baseModelId: String,
)

enum class EssentialErrorCode {
    MODEL_NOT_INSTALLED,
    MODEL_INCOMPATIBLE,
    ADAPTER_INCOMPATIBLE,
    DEVICE_CAPACITY_INSUFFICIENT,
    PERMISSION_DENIED,
    SESSION_CANCELLED,
    RUNTIME_UNAVAILABLE,
    INVALID_CONFIGURATION,
    REQUEST_TIMED_OUT,
}

class EssentialException(
    val code: EssentialErrorCode,
    override val message: String,
    cause: Throwable? = null,
) : Exception(message, cause)

internal data class ResolvedModel(
    val modelId: String?,
    val modelPath: String?,
)

internal fun nextRequestId(): String = UUID.randomUUID().toString()
